h2o-extensions/xgboost/src/test/java/hex/tree/xgboost/XGBoostTest.java

```diff
@@ -605,6 +605,68 @@ public void ProstateRegression() {
     }
   }
 
+  @Test
+  public void sparseMatrixDetectionTest() {
+    Frame tfr = null;
+    XGBoostModel model = null;
+    Scope.enter();
+    try {
+      tfr = parse_test_file("./smalldata/prostate/prostate.csv");
+      Scope.track(tfr.replace(8, tfr.vecs()[8].toCategoricalVec()));   // Convert GLEASON to categorical
+      DKV.put(tfr);
+
+      XGBoostModel.XGBoostParameters parms = new XGBoostModel.XGBoostParameters();
+      // Automatic detection should compute sparsity and decide
+      parms._dmatrix_type = XGBoostModel.XGBoostParameters.DMatrixType.auto;
+      parms._response_column = "AGE";
+      parms._train = tfr._key;
+      parms._ignored_columns = new String[]{"ID","DPROS", "DCAPS", "PSA", "VOL", "RACE", "CAPSULE"};
+
+      model = new hex.tree.xgboost.XGBoost(parms).trainModel().get();
+      assertTrue(model._output._sparse);
+
+    } finally {
+      Scope.exit();
+      if (tfr!=null) tfr.remove();
+      if (model!=null) {
+        model.delete();
+        model.deleteCrossValidationModels();
+      }
+    }
+
+  }
+
+  @Test
+  public void denseMatrixDetectionTest() {
+    Frame tfr = null;
+    XGBoostModel model = null;
+    Scope.enter();
+    try {
+      tfr = parse_test_file("./smalldata/prostate/prostate.csv");
+      DKV.put(tfr);
+
+      XGBoostModel.XGBoostParameters parms = new XGBoostModel.XGBoostParameters();
+      // Automatic detection should compute sparsity and decide
+      parms._dmatrix_type = XGBoostModel.XGBoostParameters.DMatrixType.auto;
+      parms._response_column = "AGE";
+      parms._train = tfr._key;
+      parms._ignored_columns = new String[]{"ID","DPROS", "DCAPS", "PSA", "VOL", "RACE", "CAPSULE"};
+
+      // GLEASON used as predictor variable, numeric variable, dense
+      model = new hex.tree.xgboost.XGBoost(parms).trainModel().get();
+      assertFalse(model._output._sparse);
```

Comment:
How about adding one test to detect sparsity?  Randomly generate enum columns only with high cardinality.  You can use this function: TestUtil.generate_enum_only?